flocker/control/test/test_persistence.py

```diff
@@ -249,3 +333,202 @@ def test_complex_keys(self):
+        """
+        # Get a valid v2 config.
+        v2_config = migrate_configuration(1, 2, self.v1_config, StubMigration)
+        # Perform two sequential migrations to get from v1 to v3, starting
+        # with a v1 config.
+        result = migrate_configuration(1, 3, self.v1_config, StubMigration)
+        # Compare the v1 --> v3 upgrade to the direct result of the
+        # v2 --> v3 upgrade on the v2 config, Both should be identical
+        # and valid v3 configs.
+        self.assertEqual(result, StubMigration.upgrade_from_v2(v2_config))
+
+
+UUIDS = st.basic(generate=lambda r, _: UUID(int=r.getrandbits(128)))
+
+DATASETS = st.builds(Dataset, dataset_id=UUIDS, maximum_size=st.integers())
+
+# Constrain primary to be True so that we don't get invariant errors from Node
+# due to having two differing manifestations of the same dataset id.
+MANIFESTATIONS = st.builds(
+    Manifestation, primary=st.just(True), dataset=DATASETS)
+SIMPLE_TEXT = st.text(
+    alphabet=string.letters, min_size=4, max_size=20, average_size=12
+)
+IMAGES = st.builds(DockerImage, tag=SIMPLE_TEXT, repository=SIMPLE_TEXT)
+NONE_OR_INT = st.one_of(
+    st.just(None),
+    st.integers()
+)
+ST_PORTS = st.integers(min_value=1, max_value=65535)
+PORTS = st.builds(
+    Port,
+    internal_port=ST_PORTS,
+    external_port=ST_PORTS
+)
+LINKS = st.builds(
+    Link,
+    local_port=ST_PORTS,
+    remote_port=ST_PORTS,
+    alias=SIMPLE_TEXT
+)
+APPLICATIONS = st.builds(
+    Application, name=SIMPLE_TEXT, image=IMAGES,
+    ports=st.sets(PORTS, max_size=10),
+    links=st.sets(LINKS, max_size=10),
+    environment=st.dictionaries(keys=SIMPLE_TEXT, values=SIMPLE_TEXT),
+    memory_limit=NONE_OR_INT,
+    cpu_shares=NONE_OR_INT,
+    running=st.booleans()
+)
+FILEPATHS = st.text(alphabet=string.printable).map(FilePath)
+VOLUMES = st.builds(
+    AttachedVolume, manifestation=MANIFESTATIONS, mountpoint=FILEPATHS)
+APPLICATIONS_WITH_VOLUMES = st.tuples(
+    APPLICATIONS, VOLUMES).map(lambda (a, v): a.set(volume=v))
+
+
+def _build_node(applications):
+    # All the manifestations in `applications`.
+    app_manifestations = set(app.volume.manifestation for app in applications)
+    # A set that contains all of those, plus an arbitrary set of
+    # manifestations.
+    dataset_ids = frozenset(
+        app.volume.manifestation.dataset_id
+        for app in applications if app.volume
+    )
+    manifestations = (
+        st.sets(MANIFESTATIONS.filter(
+            lambda m: m.dataset_id not in dataset_ids))
+        .map(pset)
+        .map(lambda ms: ms.union(app_manifestations))
+        .map(lambda ms: dict((m.dataset.dataset_id, m) for m in ms)))
+    return st.builds(
+        Node, uuid=UUIDS,
+        applications=st.just(applications),
+        manifestations=manifestations)
+
+
+NODES = st.sets(APPLICATIONS_WITH_VOLUMES).map(
+    lambda apps: pset(dict((
+        app.volume.manifestation.dataset_id, app) for app in apps).values())
+).flatmap(_build_node)
```

Comment:
ISTR you saying this was failing due to timeouts. I think hypothesis 1.10.3 fixes that problem. If it doesn't, we're going to have to figure out a different strategy.